Необходимо запустить практическую часть занятия- файл находиться в материалах, и посмотреть самому то, о чём говорили на лекции. По образу практики, попробуйте создать искусственный датасет с лишними столбцами. Целевую метку, при правильной обработке данных, формируйте таким образом, чтобы без затруднений её смогла описать линейная модель. Ориентируйтесь на то, что было показано во время занятия, и каждый шаг описывайте в markdown. Здесь важно видеть ваш ход мысли. Не бойтесь ошибиться или написать не то. Данное задание не имеет какого-то “правильного” решения. Цель - достичь базового понимания проблемы. Чем больше вы фантазируете, тем лучше :) Тем не менее, старайтесь представить те ситуации, которые по-вашему мнению могли бы быть в реальных данных. Успеха!



# По образу практики, попробуйте создать искусственный датасет с лишними столбцами.¶


Создадим датасет о количестве потребляемых пончиков в день:

    age - возраст (от 14 до 90)
    height - рост (от 120 до 220)
    weight - вес (от 40 до 150)
    sex - пол (M\F)
    
Будем предсказывать потербляемые пончики в зависимости от роста ,веса и возраста.

In [1]:
import numpy as np
import pandas as pd

from sklearn.linear_model import LinearRegression

Загрузили необходимые библиотеки.

In [52]:
n_samples = 600

age = np.random.choice(85, n_samples) + 5
height = np.random.choice(155, n_samples) + 55
weight = np.random.choice(80, n_samples) + 10
sex = ['F' if i < 1 else 'M' for i in np.random.choice(3, n_samples)]

donuts_ammount = height*weight/1000+7

Создаем 600 примеров, генератором случайных чисел получаем данные для этих примеров. БУдем предсказывать потребляемые пончики.

In [53]:
data = pd.DataFrame({'age': age, 'sex': sex, 'weight': weight,
                     'height': height, 'donuts_ammount': donuts_ammount})
data.head(10)

age sex  weight  height  donuts_ammount
0   20   M      62      59          10.658
1   80   M      49     130          13.370
2   31   M      76     119          16.044
3   89   F      28     137          10.836
4   50   M      77     193          21.861
5   67   M      16      60           7.960
6   17   M      85     161          20.685
7   86   F      38      79          10.002
8   58   M      52     152          14.904
9   62   M      21      57           8.197

Собственно смотрим первые 10 элементов, видим, что данные выглядят допустимым образом.

In [54]:
from sklearn.metrics import mean_absolute_error

x = data[['age', 'height', 'weight']]
y = data['donuts_ammount'] 
reg = LinearRegression().fit(x, y)
print('Weights: {}'.format(reg.coef_))
print('Bias: {}'.format(reg.intercept_))

pred_values = reg.predict(data[['age', 'height', 'weight']])
print('Error: {}'.format(mean_absolute_error(pred_values, y)))

Weights: [-0.00185814  0.04906117  0.13425592]
Bias: 0.5271037764667508
Error: 0.7717106885780933


Делаем предсказание на основании возраста, роста и веса о возрасте. Видим, что  у нас есть ошибка и мы можем что-то с  ней сделать.

In [58]:
from sklearn.metrics import mean_absolute_error

x = data[['weight', 'height']] ## плохие помыслы и вес
y = data['donuts_ammount'] ## с взрастом
reg = LinearRegression().fit(x, y)
print('Weights: {}'.format(reg.coef_))
print('Bias: {}'.format(reg.intercept_))

pred_values = reg.predict(data[['weight', 'height']])
print('Error: {}'.format(mean_absolute_error(pred_values, y)))

Weights: [0.13414437 0.04910255]
Bias: 0.43717622529766764
Error: 0.7709076370336889


Пробуем предсказать только на возраста и роста, ошибка не особо поменялась.

In [63]:
# Создаем новый признак
data['extra_weight'] = data['weight'] * data['height']/1000
data.head(5)

age sex  weight  height  donuts_ammount  extra_weight
0   20   M      62      59          10.658         3.658
1   80   M      49     130          13.370         6.370
2   31   M      76     119          16.044         9.044
3   89   F      28     137          10.836         3.836
4   50   M      77     193          21.861        14.861

Добавляем новый признак, который будет основан на предыдущих. Допустим, что набранный из-за пончиков вес зависит от веса и кол-ва пончиков.

In [64]:
X = data[['extra_weight']]
y = data['donuts_ammount']
reg = LinearRegression().fit(X, y)
print('Weights: {}'.format(reg.coef_))
print('Bias: {}'.format(reg.intercept_))

pred_values = reg.predict(data[['extra_weight']])
print('Error: {}'.format(mean_absolute_error(pred_values, y)))

Weights: [1.]
Bias: 7.000000000000002
Error: 1.791159813061919e-15


Пробуем предсказать на сонвоании этого нового признака кол-во потребляемых пончиков, видим, что ошибка у нас почти равна 0 - 1.79 в -15 степени.

# THE END!